java/memory/src/main/java/org/apache/arrow/memory/util/hash/MurmurHasher.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.memory.util.hash;
+
+import static io.netty.util.internal.PlatformDependent.getByte;
+import static io.netty.util.internal.PlatformDependent.getInt;
+
+import org.apache.arrow.memory.util.ArrowBufPointer;
+
+import io.netty.buffer.ArrowBuf;
+
+/**
+ * Implementation of the Murmur hashing algorithm.
+ * Details of the algorithm can be found in
+ * https://en.wikipedia.org/wiki/MurmurHash
+ * <p>
+ *   Murmur hashing is computationally expensive, as it involves several
+ *   integer multiplications. However, the produced hash codes have
+ *   good quality in the sense that they are uniformly distributed in the universe.
+ * </p>
+ * <p>
+ *   Therefore, this algorithm is suitable for scenarios where uniform hashing
+ *   is desired (e.g. in an open addressing hash table/hash set).
+ * </p>
+ */
+public class MurmurHasher implements ArrowBufHasher {
+
+  private final int seed;
+
+  /**
+   * Creates a default Murmur hasher, with seed 0.
+   */
+  public MurmurHasher() {
+    this(0);
+  }
+
+  /**
+   * Creates a Murmur hasher.
+   * @param seed the seed for the hasher.
+   */
+  public MurmurHasher(int seed) {
```

Comment:
not required for this PR. but, you may want to consider having the seed as an arg to the hashCode() instead of the constructor.

we do this in gandiva - it makes it possible to do chaining of hash. eg. for hash(int-vector, long-vector), we do 

```
int hash = hash32(intVec, index, 0 /*seed*/);
return hash32(longVec, index, hash /*seed*/);
```

Reply:
Thanks for the information. This looks like some useful feature. 
Maybe we can provide such APIs in a separate issue.

Reply:
ok - just pointing out that if you add an optional seed param in the API, you won't need one in the constructor.

I'm fine either way.

Reply:
Sure. Thanks for the suggestion. I have revised the code to provide a static method 

public static int hashCode(ArrowBuf buf, int offset, int length, int seed)

which should be able to support the above scenario. 
Please see if it looks good.